File: c++/623. Add One Row to Tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    TreeNode* addOneRow(TreeNode* root, int v, int d) {
        if (d==1) {
            TreeNode* newRoot = new TreeNode(v);
            newRoot->left = root;
            return newRoot;
        }
        TreeNode *p = NULL;
        queue<TreeNode*> que;
        que.push(root);

        int cnt = 0;
        while (!que.empty()) {
            int sz = que.size();
            vector<TreeNode*> vec;
            cnt++;
            while (sz--) {
                p = que.front();
                que.pop();
                vec.push_back(p);
                if (p->left)
                    que.push(p->left);
                if (p->right)
                    que.push(p->right);
            }

            if (cnt==d-1)
            {
                for (TreeNode* node : vec) {
                    TreeNode *p1 = node->left;
                    TreeNode *p2 = node->right;

                    node->left = new TreeNode(v);
                    node->right = new TreeNode(v);
                    node->left->left = p1;
                    node->right->right = p2;
                }
                break;
            }
            
        }

        return root;
    }
};

File: go/src/leetcode/623/623. Add One Row to Tree.go
/**
@author ZhengHao Lou
Date    2022/08/05
*/
package main

/**
https://leetcode.cn/problems/add-one-row-to-tree/
*/
type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

func addOneRow(root *TreeNode, val int, depth int) *TreeNode {
	var parent []*TreeNode
	que := []*TreeNode{root}
	for d := 1; d < depth && len(que) != 0; d++ {
		parent = []*TreeNode{}
		for tmp := len(que); tmp != 0; tmp-- {
			node := que[0]
			que = que[1:]
			parent = append(parent, node)
			if node.Left != nil {
				que = append(que, node.Left)
			}
			if node.Right != nil {
				que = append(que, node.Right)
			}
		}
	}

	if len(parent) == 0 {
		newRoot := &TreeNode{Val: val, Left: root}
		return newRoot
	}
	for i := range parent {
		l := parent[i].Left
		parent[i].Left = &TreeNode{Val: val, Left: l}
		r := parent[i].Right
		parent[i].Right = &TreeNode{Val: val, Right: r}
	}
	return root
}
